skimage/future/graph/_ncut.py

```diff
@@ -22,7 +22,7 @@ def DW_matrices(graph):
         joining `i` to `j`.
     """
     # sparse.eighsh is most efficient with CSC-formatted input
-    W = nx.to_scipy_sparse_matrix(graph, format='csc')
+    W = nx.to_scipy_sparse_array(graph, format='csc')
```

Comment:
How about at the top of this file adding:

```Python

if hasattr(nx, 'to_scipy_sparse_array'):
    # networkx >= 2.7
    to_scipy_sparse_array = nx.to_scipy_sparse_array
else:
    to_scipy_sparse_array = nx.to_scipy_sparse_matrix
```

and then just using that here so that it works across both versions?

Reply:
You would also need to make sure `scipy>=1.8`, but, yes, you could do that.

Reply:
I believe that would be guaranteed by the package manager. If the user has networkx 2.7 installed, tey will have scipy 1.8 installed too.

Reply:
No, not necessarily.
```
(nx1) [jarrod@x15 ~]$ pip install scipy==1.6
Collecting scipy==1.6
  Downloading scipy-1.6.0-cp39-cp39-manylinux1_x86_64.whl (27.3 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 27.3/27.3 MB 5.9 MB/s eta 0:00:00
Collecting numpy>=1.16.5
  Using cached numpy-1.22.2-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (16.8 MB)
Installing collected packages: numpy, scipy
Successfully installed numpy-1.22.2 scipy-1.6.0
(nx1) [jarrod@x15 ~]$ pip install networkx==2.7rc1
Collecting networkx==2.7rc1
  Using cached networkx-2.7rc1-py3-none-any.whl (2.0 MB)
Installing collected packages: networkx
Successfully installed networkx-2.7rc1
(nx1) [jarrod@x15 ~]$ pip list
Package    Version
---------- -------
networkx   2.7rc1
numpy      1.22.2
pip        22.0.3
scipy      1.6.0
setuptools 60.9.3
wheel      0.37.1
```

Reply:
What happens to networkx in this case? Is it usable at all? can you import it?

Reply:
Yes, it is usable; but you will have problems with anything that requires scipy.

Reply:
> You will have problems with anything that requires scipy.

To me this means it is "unusable". I don't mean to start a `pip` vs `conda` tradeoff, but to me this inconsistent environment is not one that we should consider. In my mind, pip should refuse to install that combination of software unless the user provide more explicit overrides. 

I don't think networkx 2.7 + scipy less than the requirements is something we should consider.

I do think that `networkx` is being a little aggressive in going to scipy 1.8, but that isn't really something I can control.... maybe include some shims to ensure broader compatibility?


Justification for why I think it is aggressive. Scipy 1.8 has been released less than 1 month ago.
```
$ nep29 scipy
| version |    date    |
|:-------:|:----------:|
|  1.8.0  | 2022-02-05 |
|  1.7.0  | 2021-06-20 |
|  1.6.0  | 2020-12-31 |
|  1.5.0  | 2020-06-21 |
```

Reply:
@hmaarrfk Would you checking that what you are saying about conda is true?  I would be shocked if it behaves the way you assume.

Reply:
My understanding was that scipy was a hard dependency of [networkx](https://github.com/networkx/networkx/blob/main/requirements/default.txt#L2). networkx is also a hard dependency of scikit-image.

If the recipes are correctly setup, conda would only allow a consistent environment.

For optional dependencies, there isn't a consistent syntax on conda. One way to do this (more recently) has been to use the `run_constrained` which means that if the package is not mandatory, but if it is installed, it should have specific constraints on it.

https://conda-forge.org/docs/maintainer/adding_pkgs.html?highlight=run_constrained#constraining-packages-at-runtime

The truth is, that many packages aren't always "built right" the first time. Which means often you find yourself with an old package version that doesn't have the specific lowerbound or upperbound constraints on conda. But these issues are often resolved in two ways:
1. A new package is released with updated dependencies, and the documentation can be updated to include `conda install "scikit-image>=0.20.0"` which then uses newer packages.
2. People stick with their old, but working environments.

Reply:
`scipy` is not a hard dependency of `networkx`; it is an optional dependency.  `networkx` is a pure Python package with no hard dependencies.

Reply:
Does the conda-forge NetworkX package use `run_constrained`?  Does it behave the way you expect?  I don't have conda on my computer, but if you do you could test it.

Reply:
conda and conda-forge has been more "batteries included" and actually depends on scipy:
https://github.com/conda-forge/networkx-feedstock/blob/master/recipe/meta.yaml#L24

Reply:
Interesting.  They also say we require Python 3.6, but NX 2.6.3 requires Python 3.7 or above.  It is curious they made some of our optional dependencies hard dependencies, but don't even list other optional dependencies.  Do they do any testing of their packages?

Reply:
I updated the conda-forge issue regarding our dependencies and asked them to fix the Python version issue:
https://github.com/conda-forge/networkx-feedstock/issues/1#issuecomment-1047181328

Reply:
Thank you for explaining Jarrod.

I'm confused as to what scikit-image should do in this case.

If a new networkx is installed, the two functions are largely equivelent. There is no way to avoid scipy 1.8 in either function.

i think the current "fix" as as far as we can go.

I would suggest that network manage its error messages for loose dependencies such as scipy. Maybe you can add an error message in to_scipy_sparse_array?

Reply:
Also, if anyone knows how to get Anaconda to update their packages, I've been struggling to get them to update graphviz:
https://github.com/AnacondaRecipes/graphviz-feedstock/issues/2

That version is broken and doesn't work with pygraphviz, which is one of NetworkX's optional dependencies.  Conda-forge has been more responsive and are support recent versions of graphviz:
https://anaconda.org/conda-forge/graphviz

Reply:
>  Anaconda to update their packages

You need to make a business case for Anaconda to update their packages.